### DISCUSSION TITLE
Realtimey features, aka change streams

### DISCUSSION THREAD

**Discussion by tetious** [ArcadeData/arcadedb#149] [Ideas]
I'm wondering if there are plans around or interest in implementing something like Mongo's [Change Streams](https://docs.mongodb.com/manual/changeStreams/)? 

It seems like it might not be too bad to add, given it is essentially a special-case of replication. This might be an oversimplification based on optimistic naivety, though. 😄 

Being able to subscribe to changes at the "Type" level, as well as database level, would be very interesting I think. It would also allow for fleshing out some of the more flashy bits of the Redis (pub/sub) and Mongo feature-sets.

Thoughts?

**Comment by lvca** [comment]
@tetious in OrientDB we had that: Live Queries: https://dzone.com/articles/reactive-orientdb-live-queries. @antarcticgiraffe how complex would be porting them to ArcadeDB?

On the other side, creating a proxy listener could be a quick start. Our replication log is page-based, so It can't be useful for this. The listener must be at CRUD level. We were thinking of providing such a listener mechanism anyway to hook up custom validation and plugins, so it could be a good reason to start the design of that.

**Comment by lvca** [comment]
Implemented in https://github.com/ArcadeData/arcadedb/issues/150.

**Comment by lvca** [comment]
I've added database level listener and at Type level (vertices, edges, documents).

Example of `before-record-create` listener where vertices with "validated" field equal to false cannot be saved (callback returns false):

```java
database.getEvents().registerListener((BeforeRecordCreateListener) record -> record instanceof Vertex && 
record.asVertex().getBoolean("validated"));
```

The same by only for vertex type "Client":

```java
database.getSchema().getType("Client").getEvents()
   .registerListener((BeforeRecordCreateListener) record -> record.asVertex().getBoolean("validated"));
```

**Comment by lvca** [comment]
Added new paragraph about events: https://docs.arcadedb.com/#Java-Events.

**Comment by tetious** [comment]
I like this a lot. The validation use-case is compelling!

It's been ages since I wrote any Java, but I'm thinking of taking a stab at adding a web-socket based implementation of `after-record-[create|update]` for the http API. 

My first thought for a design would be a single database-level listener that stuffs all updates/creates into a thread-safe queue. (so as not to tie up the synchronous save with network requests, or load it down with too many listeners, etc) Then a worker thread could pop and publish the record to any sockets subbed to that type/database, provided that user has read permissions for that type and/or the database as a whole. (might need to give permissions a bit more thought) 

I can think of two potential designs for the API:

1. Something like `ws://*/api/v1/listen/DATABASE/[TYPE]`, where each type gets its own connection/socket, or you leave off the type for the whole database.

If the app cares about a handful of types, but doesn't want the firehose of the whole database, that would lead to lotsa open connections, and potential connection starving, though. 

2. We could use a single endpoint, and sub/unsub via json messages sent via the socket, kinda like this: 

`ws://*/api/v1/listen`, SEND: `{"action":"[subscribe|unsubscribe]", "type": "Client", "database": "database"}`

I'm doing option 2 with a Mongo backed app right now, and it works quite well. 

Does either of these designs seem reasonable? Anything I'm missing?

EDIT: It occurs to me that this will fall apart when dealing with a cluster without some kind of backplane...

**Comment by lvca** [comment]
I like this idea and I think it could be useful to many users. ArcadeDB Server integrates JBoss Undertow, a fast and lightweight HTTP server that supports WebSocket. Maybe we could provide this out of the box?

**Reply by tetious** [reply]
That would be ideal, yep. I think it will be a good opportunity for me to dust off my rusting Java skills. Not sure of the timeline, but I'll submit a PR soonish.

My initial implementation will only work when not clustered. A cluster-friendly solution will need a backplane, but it might be possible to use ArcadeDb itself for this. I'm planning to experiment with that. First things first, though!

**Reply by tetious** [reply]
Also! Once this is properly working with clusters, I'd like to wire in support via the Redis and Mongo drivers. Should be fun!

**Reply by lvca** [reply]
Sounds good. For any help, don't hesitate to ask. Maybe Discord is quicker for this.

**Comment by lvca** [comment]
Thanks, @tetious, your PR has been merged into main and will be released with v21.11.1

**Reply by lvca** [reply]
https://github.com/ArcadeData/arcadedb/pull/163

**Comment by O-Massol** [comment]
Hello, 

We are trying to assess if we can use ArcadeDB on Remote mode to connect to a cluster of AracdeDB servers. 

At the moment we are toying with the RemoteDatabase class and the http api. 
But we are interested in listening to changes/events.
I'm browsing the discussions and issues to make sure I understand the situation and how it corresponds to my needs. 

What I understand is that the RemoteDatabase is not itself designed to subscribe to events (https://docs.arcadedb.com/#Java-Events, discussion https://github.com/ArcadeData/arcadedb/discussions/309). 
There is a websocket available (seen in this discussion, and https://docs.arcadedb.com/#http-api), but at the moment, it does not support multi-servers (or cluster). 

So my questions are : 
- Is it planned to support multi-servers websocket in the future ?
- If so, is it planned to add the events listening in the Remote driver ?
- Is there any plans to have a low-level driver (over tcp/ip for exemple) that offers the same features as the Embedded mode (regarding events and transactions) ?

Thanks in advance for your answers !

**Reply by lvca** [reply]
Since we already have the WebSocket, it could be cool to implement the Java listener API also in the remote database that uses the WebSocket under the hood. This could be a new issue.

The issue with multi-server events is our RAFT replication protocol: only modified pages are exchanged between servers. So if you update the record #10:100 on server A, its page will be sent to the other servers, but not the actual event "modified record 10:100". This is something we should add to the replication protocol.

What other features do you need in the RemoteDatabase?
